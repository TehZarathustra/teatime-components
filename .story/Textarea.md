## Preview


### size s

```
<div>
  some text for the baseline
  <Textarea
    name='textarea-s'
    size='s'
    style={{width: '200px'}}/>
  {' '}
  <Textarea
    disabled={true}
    name='textarea-s-d'
    size='s'
    style={{width: '200px'}}/>
</div>
```


### size m

```
<div>
  some text for the baseline
  <Textarea
    name='textarea-m'
    size='m'
    style={{width: '200px'}}/>
  {' '}
  <Textarea
    disabled={true}
    name='textarea-m-d'
    size='m'
    style={{width: '200px'}}/>
</div>
```
