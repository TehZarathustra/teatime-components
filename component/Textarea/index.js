/* @flow */
import {noop} from '../../lib/util';
import React, {Component} from 'react';
import classNames from 'classnames';

import textareaLarge from './textarea-large.css';
import textareaMedium from './textarea-medium.css';
import textareaSmall from './textarea-small.css';

type Props = {
  autoComplete?: 'off' | 'on',
  autoFocus: boolean,
  className?: string,
  cols?: number | string,
  defaultValue?: string,
  disabled: boolean,
  id?: string,
  maxLength?: number | string,
  name: string,
  onBlur: (event: SyntheticEvent) => void,
  onChange: (event: SyntheticInputEvent, eventData: {name: string, value: string}) => void,
  onFocus: (event: SyntheticFocusEvent) => void,
  placeholder?: string,
  readOnly: boolean,
  rows?: number | string,
  size: 'l' | 'm' | 's',
  style?: Object,
  styles?: Object,
  value?: string,
};

type TextareaProps = {
  autoComplete?: 'off' | 'on',
  autoFocus: boolean,
  className: string,
  cols?: number | string,
  defaultValue?: string,
  disabled: boolean,
  id?: string,
  maxLength?: number | string,
  name: string,
  onBlur: Function,
  onChange: (event: SyntheticInputEvent) => void,
  onFocus: Function,
  placeholder?: string,
  readOnly: boolean,
  rows?: number | string,
  style?: Object,
  value?: string,
};

const cssModules = {
  l: textareaLarge,
  m: textareaMedium,
  s: textareaSmall,
};

class Textarea extends Component {
  props: Props

  static defaultProps: {
    autoFocus: boolean,
    disabled: boolean,
    onBlur: Function,
    onChange: Function,
    onFocus: Function,
    readOnly: boolean,
    size: 'l' | 'm' | 's',
  }

  css: {
    textarea: string,
  }

  _textareaElement: HTMLTextAreaElement

  constructor(props: Props) {
    super(props);

    this.css = props.styles || cssModules[props.size];

    (this:any)._onChange = this._onChange.bind(this);
    (this:any)._saveRef = this._saveRef.bind(this);
  }

  componentWillReceiveProps(nextProps: Props) {
    this.css = nextProps.styles || cssModules[nextProps.size];
  }

  focus() {
    if (this._textareaElement) this._textareaElement.focus();
  }

  select() {
    if (this._textareaElement) this._textareaElement.select();
  }

  _computeTextareaProps(): TextareaProps {
    const {
      autoComplete,
      autoFocus,
      className,
      cols,
      defaultValue,
      disabled,
      id,
      maxLength,
      name,
      onBlur,
      onFocus,
      placeholder,
      readOnly,
      rows,
      style,
      value,
    } = this.props;
    const {css} = this;

    return {
      autoComplete,
      autoFocus,
      className: classNames(css.textarea, className),
      cols,
      defaultValue: defaultValue !== null ? defaultValue : '',
      disabled,
      id,
      maxLength,
      name,
      onBlur,
      onChange: this._onChange,
      onFocus,
      placeholder,
      readOnly,
      rows,
      style,
      value: value !== null ? value : '',
    };
  }

  _onChange(e: SyntheticInputEvent) {
    const {name, onChange} = this.props;
    onChange(e, {name, value: e.target.value});
  }

  _saveRef(ref: HTMLTextAreaElement) {
    this._textareaElement = ref;
  }

  render() {
    const textareaProps = this._computeTextareaProps();

    return (
      <textarea {...textareaProps}/>
    );
  }
}

Textarea.defaultProps = {
  autoFocus: false,
  disabled: false,
  onBlur: noop,
  onChange: noop,
  onFocus: noop,
  readOnly: false,
  size: 'm',
};

export default Textarea;
