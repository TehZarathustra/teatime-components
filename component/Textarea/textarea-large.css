.textarea
{
  composes: control from '../Input/input.css';

  padding: 6px 9px;

  font: 15px/1.2 arial,sans-serif;
}
