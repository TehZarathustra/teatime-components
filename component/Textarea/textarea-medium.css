.textarea
{
  composes: control from '../Input/input.css';

  padding: 5px 7px;

  font: 13px/1.2 arial,sans-serif;
}
