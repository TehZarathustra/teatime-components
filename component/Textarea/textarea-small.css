.textarea
{
  composes: control from '../Input/input.css';

  padding: 3px 6px;

  font: 13px/1.2 arial,sans-serif;
}
