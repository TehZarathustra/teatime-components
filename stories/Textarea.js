'use strict';

const { action, storiesOf } = require('@kadira/storybook');
const { column, columns, delimiter } = require('./stories.css');
const React = require('react');
const Textarea = require('../component/Textarea');

const labelStyles = {
  display: 'inline-block',
  margin: '0 8px 0 0',
  width: '62px',
};
const inputStyles = {
  width: 'calc(100% - 70px)',
};

storiesOf('Textarea', module)
  .add('preview', () => (
    <div className={columns}>
      <div className={column}>
        <span style={labelStyles}>size: `s`</span>
        <Textarea
          name='es'
          onChange={action('change')}
          size='s'
          style={inputStyles}/>

        <hr className={delimiter}/>

        <span style={labelStyles}>size: `m`</span>
        <Textarea
          name='em'
          onChange={action('change')}
          size='m'
          style={inputStyles}/>
      </div>
      <div className={column}>
        <Textarea
          disabled={true}
          name='ds'
          onChange={action('change')}
          size='s'/>

        <hr className={delimiter}/>

        <Textarea
          disabled={true}
          name='dm'
          onChange={action('change')}
          size='m'/>
      </div>
    </div>
  ));
