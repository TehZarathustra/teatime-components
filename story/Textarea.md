```
<div className='columns'>
  <div className='column'>
    <span className='label'>size `s`</span>
    <Textarea className='control' name='es' size='s'/>

    <hr className='delimiter'/>

    <span className='label'>size `m`</span>
    <Textarea className='control' name='em' size='m'/>
  </div>
  <div className='column'>
    <Textarea disabled={true} name='ds' size='s'/>

    <hr className='delimiter'/>

    <Textarea disabled={true} name='dm' size='m'/>
  </div>
</div>
```
