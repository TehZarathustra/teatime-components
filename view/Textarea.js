'use strict';

const React = require('react');
const { Component, PropTypes } = React;
const { bind, noop } = require('../tools/func');
const cssModules = require('react-css-modules');

class Textarea extends Component {
  constructor(props) {
    super(props);

    bind(this, 'onChange');
  }

  onChange(e) {
    this.props.onChange(e, {value: e.target.value});
  }

  render() {
    return (
      <textarea styleName='control' {...this.props} onChange={this.onChange}/>
    );
  }
}

Textarea.defaultProps = {
  onChange: noop,
  rows: 3,
  styles: {},
};

Textarea.propTypes = {
  cols: PropTypes.number,
  defaultValue: PropTypes.string,
  disabled: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onCopy: PropTypes.func,
  onCut: PropTypes.func,
  onFocus: PropTypes.func,
  onInput: PropTypes.func,
  onKeyDown: PropTypes.func,
  onKeyPress: PropTypes.func,
  onKeyUp: PropTypes.func,
  onPaste: PropTypes.func,
  onScroll: PropTypes.func,
  onSelect: PropTypes.func,
  onWheel: PropTypes.func,
  placeholder: PropTypes.string,
  rows: PropTypes.number,
  styles: PropTypes.object,
  value: PropTypes.string,
};

module.exports = cssModules(Textarea);
