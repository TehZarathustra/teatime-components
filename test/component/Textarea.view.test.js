'use strict';

import React from 'react';
import Textarea from '../../component/Textarea';
import renderer from 'react-test-renderer';

test('defaultProp #size should have a value `m`', () => {
  expect(Textarea.defaultProps.size).toBe('m');
});

test('Renders correctly with default props', () => {
  const tree = renderer.create(
    <Textarea
      name='control'/>
  );

  expect(tree).toMatchSnapshot();
});

test('Renders correctly with defined props', () => {
  const tree = renderer.create(
    <Textarea
      autoComplete='off'
      autoFocus
      className='outer'
      cols='5'
      disabled
      id='controls-id'
      maxLength='25'
      name='control'
      placeholder='it is a placeholder'
      readOnly
      rows='10'
      value='Any Code...'/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
