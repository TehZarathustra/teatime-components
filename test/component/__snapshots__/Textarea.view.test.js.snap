// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with default props 1`] = `
<textarea
  autoComplete={undefined}
  autoFocus={false}
  className="textarea"
  cols={undefined}
  defaultValue={undefined}
  disabled={false}
  id={undefined}
  maxLength={undefined}
  name="control"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder={undefined}
  readOnly={false}
  rows={undefined}
  style={undefined}
  value={undefined}
/>
`;

exports[`Renders correctly with defined props 1`] = `
<textarea
  autoComplete="off"
  autoFocus={true}
  className="textarea outer"
  cols="5"
  defaultValue={undefined}
  disabled={true}
  id="controls-id"
  maxLength="25"
  name="control"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder="it is a placeholder"
  readOnly={true}
  rows="10"
  style={undefined}
  value="Any Code..."
/>
`;
